### PR TITLE
Expanding mtime to file watches

### PR DESCRIPTION
**Pull request by nyarly**
This is fairly big PR, but it was hard to limit the scope - it's kind of a fundamental change but a really nice one. It would be rad if this got merged as is, but I think that's beyond expectation. If there's changes needed, it'd be nice to hear as soon as possible so that it's still fresh in the meat caches.

I've swapped out the mtime in the RC struct for a FileTimes list that records the path and mtime for mutliple files. The watch structure is serialized to DIRENV_WATCHES the same way DIRENV_DIFF is.

The upshot is that direnv can now watch more than just the .envrc and its allow file for changes. This addresses #173, #98 as well as #77.

## Summary of Changes

Obviously, there's the FileTime[s] structs and associated methods. There were a lot of changes to `direnv export` and `direnv edit` to handle the change is how things were tracked.

### New Private Commands
`direnv watch <path>` adds the path to the watchlist and prints the new export command for the DIRENV_WATCHES env var. The upshot is that stdlib and direnvrc functions can `eval $(direnv watch filename)` in order to add a file to the list of watches. (export captures the value at the end of loading the .envrc so that changes during its execution can be picked up)

`direnv current <path>` checks to see if <path> is current or not - it exits 0 if the path is current or 1 if it's unknown, stale, missing or newly added. Almost useless outside of .envrc (or .direnvrc), but could be very useful for writing a caching stdlib function.

## Mea Culpas
A few changes not strictly in scope:

The behavior of the watch stuff was tricky to debug, so I wound up extending direnv's debugging - log.go now has a log_debug function that is a no-op unless `DIRENV_DEBUG=1` is set. It makes use of the log package instead of Fprintf'ing to Stderr, which is nice because it makes it possible to track file and line number.

cmd_export.go got a heavy refactoring. I had a hard time reasoning about it in one giant method, which is why it got broken up like that.

Maybe the worst thing: I updated the version to 2.8 - I needed an easy way to make sure I was running the right version when I was taking it for a test run, but I don't think that's exactly kosher. Still, the change of interface requires at least that, I think?

### CODE REVIEW

**Comment by zimbatm** [comment]
Love this. I will take a look at the code now.

File: env.go [hunk 1]
```diff
@@ -25,6 +25,12 @@ func GetEnv() Env {
 	return env
 }
 
+func (env Env) Clean() {
```
Comment (zimbatm) [inline env.go]: Maybe rename to `CleanState()` or `CleanContext()`

File: stdlib.sh [hunk 2]
```diff
@@ -153,7 +153,10 @@ source_env() {
   if ! [[ -f $rcpath ]]; then
     rcpath=$rcpath/.envrc
   fi
+
   rcfile=$(user_rel_path "$rcpath")
+  watch_file $rcpath
```
Comment (zimbatm) [inline stdlib.sh]: the `"$rcpath"` should be quoted in case it contains some spaces

File: stdlib.sh [hunk 3]
```diff
@@ -166,6 +169,18 @@ source_env() {
...
 
+# Usage: watch_file <filename>
+#
+# Adds <path> to the list of files that direnv will watch for changes - useful when the contents
+# of a file influence how variables are set - especially in direnvrc
+#
+watch_file() {
+  local file=${1/#\~/$HOME}
+
+  eval $($direnv watch $file)
```
Comment (zimbatm) [inline stdlib.sh]: quoted `"$file"`.

Shell escaping is tricky. I always use shellcheck on all of my scripts because of that.

File: stdlib.sh [hunk 4]
```diff
@@ -486,7 +501,7 @@ use_nix() {
 
 ## Load the global ~/.direnvrc if present
 if [[ -f ${XDG_CONFIG_HOME:-$HOME/.config}/direnv/direnvrc ]]; then
-  source_env "${XDG_CONFIG_HOME:-$HOME/.config}/direnv/direnvrc" >&2
+  source_env ${XDG_CONFIG_HOME:-$HOME/.config}/direnv/direnvrc >&2
 elif [[ -f $HOME/.direnvrc ]]; then
-  source_env "$HOME/.direnvrc" >&2
+  source_env $HOME/.direnvrc >&2
```
Comment (zimbatm) [inline stdlib.sh]: I'm not sure why the quotes were removed here. On OSX the home directly can be "/Users/Foo Bar" I think

File: version.go [hunk 5]
```diff
@@ -1,3 +1,3 @@
 package main
 
-const VERSION = "2.7.0"
+const VERSION = "2.8.0"
```
Comment (zimbatm) [inline version.go]: I agree but I prefer to do the release separately. I like to bump the version and update the CHANGELOG.md at the same time (and then tag that commit).

**Comment by zimbatm** [comment]
:tada: Got a few comments above ^ and Travis might have to be kicked around a bit. I suspect that we lost go 1.4 compatibility but that's not a bit deal. The docs also need updating a bit but I'm happy to do that independently.

Overall I really like the idea and the implementation is great. The export function is the hairiest and I like what you did with it. I also like the storage of mtimes and the overall cleanup that you did.

If you wanted to take things even further, all the state could be stored in a single `DIRENV_STATE` object that would contain the dir, diff and mtimes. That would open the door to adding other features like multiple switchable context per folder (for example to test a library with different versions of ruby).

**Comment by nyarly** [comment]
Updated - ran shellcheck and ensured that all the checks are cleared, including setting some source= directives; has the advantage of hinting at shellcheck for other devs.

**Comment by zimbatm** [comment]
Looking good. Anything more ?

**Comment by nyarly** [comment]
I think that's it. I like the idea of a unified STATE but... One think at a
time, eh?

**Comment by zimbatm** [comment]
Sure thing :)

I'm going to do some more testing, adding the docs and also make use of the new feature in `source_env`, `dotenv`, ... in the next couple of days. Once I'm happy that it's stable I'll make a release, probably next weekend.

**Pull request merged**